### PR TITLE
Add support for multiple result node

### PR DESCRIPTION
**Pull request by soxofaan**
Issue #391

### CODE REVIEW

**Comment by soxofaan** [comment]
todo
- [ ] changelog entry
- [ ] docs
- [ ] more tests?

File: openeo/rest/_testing.py [hunk 1]
```diff
@@ -69,6 +81,35 @@ def __init__(self, requests_mock, connection: Connection):
         )
         requests_mock.post(connection.build_url("/validation"), json=self._handle_post_validation)
 
+    @classmethod
+    def at(cls, root_url: str, *, requests_mock, capabilities: Optional[dict] = None) -> DummyBackend:
```
Comment (soxofaan) [inline openeo/rest/_testing.py:85]: rename to `from_url` to be less cryptic?

File: openeo/rest/_testing.py [hunk 2]
```diff
@@ -147,10 +188,21 @@ def get_sync_pg(self) -> dict:
...
+        Get process graph of the one and only batch job.
+        Fails when there is none or more than one.
+        """
         assert len(self.batch_jobs) == 1
         return self.batch_jobs[max(self.batch_jobs.keys())]["pg"]
 
+    def get_validation_pg(self) -> dict:
+        """
+        Get process graph of the one and only validation request.
+        :return:
```
Comment (soxofaan) [inline openeo/rest/_testing.py:201]: remove/fix this useless line

File: openeo/rest/multiresult.py [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+
+from typing import Dict, List, Optional
+
+from openeo import BatchJob
+from openeo.internal.graph_building import FlatGraphableMixin, MultiLeafGraph
+from openeo.rest import OpenEoClientException
+from openeo.rest.connection import Connection, extract_connections
+
+
+class MultiResult(FlatGraphableMixin):
```
Comment (soxofaan) [inline openeo/rest/multiresult.py:11]: Include this in API docs

File: openeo/rest/multiresult.py [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+        cube2 = ...
+        multi_result = MultiResult([cube1, cube2])
+        job = multi_result.create_job()
+
+
+    """
+
+    def __init__(self, leaves: List[FlatGraphableMixin], connection: Optional[Connection] = None):
+        self._multi_leaf_graph = MultiLeafGraph(leaves=leaves)
+        self._connection = self._common_connection(leaves=leaves, connection=connection)
```
Comment (soxofaan) [inline openeo/rest/multiresult.py:29]: TODO: fail on finding common connection should be lazy (when user actually triggers need for a connection)

File: tests/rest/test_connection.py [hunk 5]
```diff
@@ -52,6 +53,7 @@
 refresh_token_store = refresh_token_store
 
 
+
```
Comment (soxofaan) [inline tests/rest/test_connection.py:56]: stray whitespace

File: openeo/rest/_testing.py (hunk 1 above)
Comment (soxofaan) [reply openeo/rest/_testing.py:85]: or `at_url`

File: openeo/rest/multiresult.py (hunk 4 above)
Comment (soxofaan) [reply openeo/rest/multiresult.py:50]: won't fix: MultiResult is really connection-centric and should "fail fast" here. if connection-less flattening is desired, `MultiLeafGraph` can be used

File: openeo/internal/graph_building.py [hunk 6]
```diff
@@ -438,3 +451,24 @@ def _process_from_parameter(self, name: str) -> Any:
         if name not in self._parameters:
             raise ProcessGraphVisitException("No substitution value for parameter {p!r}.".format(p=name))
         return self._parameters[name]
+
+
+class MultiLeafGraph(FlatGraphableMixin):
+    """
+    Container for process graphs with multiple leaf/result nodes.
+    """
+
```
Comment (soxofaan) [inline openeo/internal/graph_building.py:460]: make slotted

File: openeo/rest/multiresult.py [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+        multi_result = MultiResult([cube1, cube2])
+        job = multi_result.create_job()
+
+    .. seealso::
+
+        :ref:`multi-result-process-graphs`
+
+    .. versionadded:: 0.35.0
+    """
+
```
Comment (soxofaan) [inline openeo/rest/multiresult.py:36]: make slotted

File: tests/rest/test_multiresult.py [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+            "saveresult1": {
+                "process_id": "save_result",
+                "arguments": {"data": {"from_node": "loadcollection1"}, "format": "GTiff", "options": {}},
+            },
+            "saveresult2": {
+                "process_id": "save_result",
+                "arguments": {"data": {"from_node": "loadcollection1"}, "format": "netCDF", "options": {}},
+                "result": True,
+            },
+        }
```
Comment (soxofaan) [inline tests/rest/test_multiresult.py:53]: add tests for flat_graph and execute_batch too

**Comment by soxofaan** [comment]
I think this is ready to merge, but still a quick review could be handy to make sure  there is no mismatch in expectations. @jdries  @HansVRP  @VincentVerelst 

This is quite a large PR in terms of lines of code, but the essential tl;dr is  basically described at https://github.com/Open-EO/openeo-python-client/pull/651/files#diff-28f4328e7d44f5afd6d4a6ce10ed344de812d2f2d9460a913c8e65e0a77cdeb5

**Review by jdries — APPROVED** [review]

**Comment by jdries** [comment]
approved, one general comment about imports: for our public API, I would prefer to import like:
`from openeo import MultiResult`
This also works for DataCube for instance, but was apparently also not done for VectorCube. Why do we push the visibility of the 'rest' module?

**Comment by soxofaan** [comment]
> from openeo import MultiResult

good point, added

**Pull request merged**